### PR TITLE
[Chat Input ] Edit message cursor jump fix

### PR DESCRIPTION
**Pull request by gtalha07**
- Fixes[ #meta-issue-468](https://github.com/acterglobal/a3-meta/issues/468). The changes include removing ```message``` string we were storing in ```chatInputProvider``` for tracking send button updates. This combination with ```TextEditingController``` is problematic as what've seen from the past where the states aren't properly synced in between.  The ```ValueListener``` over controller to listen controller changes directly, provides more robustness to send button state updates. Another side-effect of riverpod  with ```TextEditingController``` is cursor jumps when editing message in selection .i.e. controller state resets upon riverpod ```updateMessage()``` calls which also shifts cursor to the end of selection.

https://github.com/user-attachments/assets/9f8c222c-a2f1-479f-8788-17d60792b3e4

### CODE REVIEW

**Comment by gnunicorn** [comment]
@gtalha07 in the video the top bar (most notably the user count) flickers. why is that? is that new?

File: app/test/features/chat/custom_input.dart [hunk 1]
```diff
@@ -170,19 +216,21 @@ void main() {
       (tester) async {
         await tester.pumpWidget(
           ProviderScope(
-            overrides: overrides,
+            overrides: [
+              sdkProvider.overrideWith((ref) => MockActerSdk()),
+              ...overrides,
```
Comment (gnunicorn) [inline app/test/features/chat/custom_input.dart:221]: nice! but do we really need to mock the ActerSDK itself? Isn't it enough to mock a higher level provider instead?

File: app/test/features/chat/custom_input.dart [hunk 2]
```diff
@@ -212,10 +260,60 @@ void main() {
         await tester.pump();
         await tester.enterTextWithoutReplace(find.byType(TextField), 't');
 
-        await tester.pump();
+        // This test is timing out due to a pending timer.
+        // See MultiTriggerAutocompleteState._onChangedField in:
+        // acter_trigger_autocomplete.dart:279
+        // put 300ms delay as (debounceTimerDuration)
+        await tester.pump(Durations.medium2);
         expect(controller.text, 'testing code'); // <- cursor moved
```
Comment (gnunicorn) [inline app/test/features/chat/custom_input.dart:268]:
```suggestion
        expect(controller.text, 'testing code');
```

File: app/test/features/chat/custom_input.dart [hunk 3]
```diff
@@ -212,10 +260,60 @@ void main() {
...
+
+        final element = tester.element(find.byType(CustomChatInput));
+        final container = ProviderScope.containerOf(element);
+
+        final TextField textField = tester.widget(find.byType(TextField));
+        final controller = textField.controller!;
+
+        // initial state should be empty
+        assert(controller.text.trim().isEmpty, true);
+
```
Comment (gnunicorn) [inline app/test/features/chat/custom_input.dart:300]: Can I also has a test that confirms if it wasn't empty and we route to another chat it is not the same and come back, it holds the previous state for that same room? That's the main concern I have with this changeset and a test proofing it works would relieve my worry.

File: app/lib/features/chat/widgets/custom_input.dart [hunk 4]
```diff
@@ -692,10 +705,23 @@ class _TextInputWidgetConsumerState extends ConsumerState<_TextInputWidget> {
...
+        // frame delay to keep focus connected with keyboard.
+        WidgetsBinding.instance.addPostFrameCallback((_) {
+          widget.chatFocus.requestFocus();
+        });
+      } else if (next == SelectedMessageState.replyTo) {
+        // frame delay to keep focus connected with keyboard..
+        WidgetsBinding.instance.addPostFrameCallback((_) {
+          widget.chatFocus.requestFocus();
+        });
+      }
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/custom_input.dart:724]: isn't it easier to just wrap the entire part in the PostFrameCallback? It's also more legible...
```suggestion
        WidgetsBinding.instance.addPostFrameCallback((_) {
      if (next == SelectedMessageState.edit) {
        // a new message has been selected to be edited or switched from reply
        // to edit, force refresh the inner text controller to reflect that
        widget.controller.text = parseEditMsg(ref);
        // frame delay to keep focus connected with keyboard.
          widget.chatFocus.requestFocus();
      } else if (next == SelectedMessageState.replyTo) {
        // frame delay to keep focus connected with keyboard..
          widget.chatFocus.requestFocus();
      }
        });
```

File: app/lib/features/chat/widgets/custom_input.dart [hunk 5]
```diff
@@ -199,11 +194,19 @@ class _ChatInput extends ConsumerStatefulWidget {
 class __ChatInputState extends ConsumerState<_ChatInput> {
   late ActerTriggerAutoCompleteTextController textController;
   final FocusNode chatFocus = FocusNode();
+  final ValueNotifier<bool> _isInputEmptyNotifier = ValueNotifier(true);
 
   @override
   void didChangeDependencies() {
     super.didChangeDependencies();
     _setController();
+    textController.addListener(_updateInputState);
```
Comment (gnunicorn) [inline app/lib/features/chat/widgets/custom_input.dart:203]: shouldn't we do this within `_setController`?

File: app/lib/features/chat/utils.dart [hunk 6]
```diff
@@ -244,3 +246,32 @@ String prepareMsg(MsgContent? content) {
     (match) => '<a href="${match.group(0)}">${match.group(0)}</a>',
   );
 }
+
+String parseEditMsg(WidgetRef ref) {
+  // shouldn't ever happen to be null if state is edit or reply
+  final message = ref.read(chatInputProvider).selectedMessage!;
```
Comment (gnunicorn) [inline app/lib/features/chat/utils.dart:252]: may I suggest a change of API here...

```suggestion
String parseEditMsg(Message message) {
```

and also, as this is then nicely easily input-output-based ... can we have unit tests for this confirming it does what we want? Should be rather straight forward then.

**Review by gnunicorn — APPROVED** [review]
\o/ tests and fixing them!

Just a few minor requests to changes in the code. Most concerned I am about adding one more test about re-entry. If that can be added and proven, go ahead and merge this :) .

**Comment by gtalha07** [comment]
> @gtalha07 in the video the top bar (most notably the user count) flickers when you type. why is that? is that new?

It's because of typing notice call which goes when user starts typing on field. This has been as it is before.

File: app/test/features/chat/custom_input.dart (hunk 1 above)
Comment (gtalha07) [reply app/test/features/chat/custom_input.dart:221]: The log points out to the last trace of sdk which is chained with other providers in the tree. The lib missing error only seems to resolve with this override not the other ones...

File: app/test/features/chat/custom_input.dart (hunk 3 above)
Comment (gtalha07) [reply app/test/features/chat/custom_input.dart:300]: The new refactored mentions code (with text controller) isn't storing any previous state of room when switching. As I wanted to switch this mechanism with ```compose draft``` in the follower PRs. Do you think we want to switch it back to riverpod version of state storing for now? And migrate that to ```composeDraft``` in separate PR ? OR wait ? I will test in both cases...

File: app/lib/features/chat/widgets/custom_input.dart (hunk 4 above)
Comment (gtalha07) [reply app/lib/features/chat/widgets/custom_input.dart:724]: It can be. Thanks for pointer!

Comment (gtalha07) [reply app/lib/features/chat/widgets/custom_input.dart:724]: Unfortunately, it throws error because ```ref``` update call. We have to keep it as before as ref isn't valid in next frame...
```
════════ Exception caught by scheduler library ═════════════════════════════════
Bad state: Cannot use "ref" after the widget was disposed.

```

File: app/test/features/chat/custom_input.dart (hunk 3 above)
Comment (gnunicorn) [reply app/test/features/chat/custom_input.dart:300]: well, could you add a test and have it skipped for now? and we'll fix it in the composedraft one? but then this second one needs to come soon .... before the next release ideally ...
